File: ios/Kuulla/Kuulla/SettingsClient.swift
import Foundation

struct SettingsClient {
    private let apiClient: ApiClient

    init(apiClient: ApiClient = .shared) {
        self.apiClient = apiClient
    }

    func getSettings() async throws -> UserSettings {
        try await apiClient.get(["api", "settings"])
    }

    func updateUnlistenedEpisodeCount(_ value: UnlistenedEpisodeCount) async throws -> UserSettings {
        try await apiClient.put(["api", "settings"], body: UpdateSettingsRequest(unlistenedEpisodeCount: value))
    }

    func updateSubscriptionSortOrder(_ value: SubscriptionSortOrder) async throws -> UserSettings {
        try await apiClient.put(
            ["api", "settings", "subscription-sort-order"],
            body: UpdateSubscriptionSortOrderRequest(subscriptionSortOrder: value))
    }

    func updateSubscriptionManualOrder(_ showIds: [String]) async throws -> UserSettings {
        try await apiClient.put(
            ["api", "settings", "subscription-manual-order"],
            body: UpdateSubscriptionManualOrderRequest(showIds: showIds))
    }

    func updateHideCaughtUpShows(_ value: Bool) async throws -> UserSettings {
        try await apiClient.put(
            ["api", "settings", "hide-caught-up-shows"],
            body: UpdateHideCaughtUpShowsRequest(hideCaughtUpShows: value))
    }

    func getShowSettings(showId: String) async throws -> ShowSettings {
        try await apiClient.get(["api", "settings", "shows", showId])
    }

    func updateShowUnlistenedEpisodeCount(showId: String, value: UnlistenedEpisodeCount?) async throws -> ShowSettings {
        try await apiClient.put(
            ["api", "settings", "shows", showId],
            body: UpdateShowSettingsRequest(unlistenedEpisodeCount: value))
    }

    func updateAutoArchiveRule(_ value: AutoArchiveRule) async throws -> UserSettings {
        try await apiClient.put(["api", "settings", "auto-archive"], body: UpdateAutoArchiveRuleRequest(autoArchiveRule: value))
    }

    func updateShowAutoArchiveRule(showId: String, value: AutoArchiveRule?) async throws -> ShowSettings {
        try await apiClient.put(
            ["api", "settings", "shows", showId, "auto-archive"],
            body: UpdateShowAutoArchiveRuleRequest(autoArchiveRule: value))
    }

    func updateAutoSkip(introSeconds: Int, outroSeconds: Int) async throws -> UserSettings {
        try await apiClient.put(
            ["api", "settings", "auto-skip"],
            body: UpdateAutoSkipRequest(autoSkipIntroSeconds: introSeconds, autoSkipOutroSeconds: outroSeconds))
    }

    func updateShowAutoSkip(showId: String, introSeconds: Int?, outroSeconds: Int?) async throws -> ShowSettings {
        try await apiClient.put(
            ["api", "settings", "shows", showId, "auto-skip"],
            body: UpdateShowAutoSkipRequest(autoSkipIntroSeconds: introSeconds, autoSkipOutroSeconds: outroSeconds))
    }

    func updatePlaybackSpeed(_ value: Float) async throws -> UserSettings {
        try await apiClient.put(["api", "settings", "playback-speed"], body: UpdatePlaybackSpeedRequest(playbackSpeed: value))
    }

    func updateShowPlaybackSpeed(showId: String, value: Float?) async throws -> ShowSettings {
        try await apiClient.put(
            ["api", "settings", "shows", showId, "playback-speed"],
            body: UpdateShowPlaybackSpeedRequest(playbackSpeed: value))
    }

    func updateAutoDeleteRule(_ rule: AutoDeleteRule, afterDays: Int) async throws -> UserSettings {
        try await apiClient.put(
            ["api", "settings", "auto-delete"],
            body: UpdateAutoDeleteRuleRequest(autoDeleteRule: rule, autoDeleteAfterDays: afterDays))
    }

    func updateShowAutoDeleteRule(showId: String, rule: AutoDeleteRule?, afterDays: Int?) async throws -> ShowSettings {
        try await apiClient.put(
            ["api", "settings", "shows", showId, "auto-delete"],
            body: UpdateShowAutoDeleteRuleRequest(autoDeleteRule: rule, autoDeleteAfterDays: afterDays))
    }

    func updateAutoDownloadNewEpisodes(_ value: Bool) async throws -> UserSettings {
        try await apiClient.put(
            ["api", "settings", "auto-download"],
            body: UpdateAutoDownloadNewEpisodesRequest(autoDownloadNewEpisodes: value))
    }

    func updateShowAutoDownloadNewEpisodes(showId: String, value: Bool?) async throws -> ShowSettings {
        try await apiClient.put(
            ["api", "settings", "shows", showId, "auto-download"],
            body: UpdateShowAutoDownloadNewEpisodesRequest(autoDownloadNewEpisodes: value))
    }

    func updateAutoAddNewEpisodesToUpNext(_ value: Bool) async throws -> UserSettings {
        try await apiClient.put(
            ["api", "settings", "auto-add-up-next"],
            body: UpdateAutoAddNewEpisodesToUpNextRequest(autoAddNewEpisodesToUpNext: value))
    }

    func updateShowAutoAddNewEpisodesToUpNext(showId: String, value: Bool?) async throws -> ShowSettings {
        try await apiClient.put(
            ["api", "settings", "shows", showId, "auto-add-up-next"],
            body: UpdateShowAutoAddNewEpisodesToUpNextRequest(autoAddNewEpisodesToUpNext: value))
    }

    func updateUpNextInsertPosition(_ value: UpNextInsertPosition) async throws -> UserSettings {
        try await apiClient.put(
            ["api", "settings", "up-next-insert-position"],
            body: UpdateUpNextInsertPositionRequest(upNextInsertPosition: value))
    }

    func updateSmartSpeed(_ value: Bool) async throws -> UserSettings {
        try await apiClient.put(["api", "settings", "smart-speed"], body: UpdateSmartSpeedRequest(smartSpeed: value))
    }

    func updateShowSmartSpeed(showId: String, value: Bool?) async throws -> ShowSettings {
        try await apiClient.put(
            ["api", "settings", "shows", showId, "smart-speed"],
            body: UpdateShowSmartSpeedRequest(smartSpeed: value))
    }

    func updateNotificationsEnabled(_ value: Bool) async throws -> UserSettings {
        try await apiClient.put(
            ["api", "settings", "notifications"],
            body: UpdateNotificationsEnabledRequest(notificationsEnabled: value))
    }

    func updateShowNotificationsEnabled(showId: String, value: Bool?) async throws -> ShowSettings {
        try await apiClient.put(
            ["api", "settings", "shows", showId, "notifications"],
            body: UpdateShowNotificationsEnabledRequest(notificationsEnabled: value))
    }

    func updateSleepTimerDefaultDuration(_ minutes: Int) async throws -> UserSettings {
        try await apiClient.put(
            ["api", "settings", "sleep-timer-default-duration"],
            body: UpdateSleepTimerDefaultDurationRequest(sleepTimerDefaultDurationMinutes: minutes))
    }
}

private struct UpdateSettingsRequest: Encodable {
    let unlistenedEpisodeCount: UnlistenedEpisodeCount
}

private struct UpdateSubscriptionSortOrderRequest: Encodable {
    let subscriptionSortOrder: SubscriptionSortOrder
}

private struct UpdateSubscriptionManualOrderRequest: Encodable {
    let showIds: [String]
}

private struct UpdateHideCaughtUpShowsRequest: Encodable {
    let hideCaughtUpShows: Bool
}

private struct UpdateShowSettingsRequest: Encodable {
    let unlistenedEpisodeCount: UnlistenedEpisodeCount?
}

private struct UpdateAutoArchiveRuleRequest: Encodable {
    let autoArchiveRule: AutoArchiveRule
}

private struct UpdateShowAutoArchiveRuleRequest: Encodable {
    let autoArchiveRule: AutoArchiveRule?
}

private struct UpdateAutoSkipRequest: Encodable {
    let autoSkipIntroSeconds: Int
    let autoSkipOutroSeconds: Int
}

private struct UpdateShowAutoSkipRequest: Encodable {
    let autoSkipIntroSeconds: Int?
    let autoSkipOutroSeconds: Int?
}

private struct UpdatePlaybackSpeedRequest: Encodable {
    let playbackSpeed: Float
}

private struct UpdateShowPlaybackSpeedRequest: Encodable {
    let playbackSpeed: Float?
}

private struct UpdateAutoDeleteRuleRequest: Encodable {
    let autoDeleteRule: AutoDeleteRule
    let autoDeleteAfterDays: Int
}

private struct UpdateShowAutoDeleteRuleRequest: Encodable {
    let autoDeleteRule: AutoDeleteRule?
    let autoDeleteAfterDays: Int?
}

private struct UpdateAutoDownloadNewEpisodesRequest: Encodable {
    let autoDownloadNewEpisodes: Bool
}

private struct UpdateShowAutoDownloadNewEpisodesRequest: Encodable {
    let autoDownloadNewEpisodes: Bool?
}

private struct UpdateAutoAddNewEpisodesToUpNextRequest: Encodable {
    let autoAddNewEpisodesToUpNext: Bool
}

private struct UpdateShowAutoAddNewEpisodesToUpNextRequest: Encodable {
    let autoAddNewEpisodesToUpNext: Bool?
}

private struct UpdateUpNextInsertPositionRequest: Encodable {
    let upNextInsertPosition: UpNextInsertPosition
}

private struct UpdateSmartSpeedRequest: Encodable {
    let smartSpeed: Bool
}

private struct UpdateShowSmartSpeedRequest: Encodable {
    let smartSpeed: Bool?
}

private struct UpdateNotificationsEnabledRequest: Encodable {
    let notificationsEnabled: Bool
}

private struct UpdateShowNotificationsEnabledRequest: Encodable {
    let notificationsEnabled: Bool?
}

private struct UpdateSleepTimerDefaultDurationRequest: Encodable {
    let sleepTimerDefaultDurationMinutes: Int
}

File: src/Kuulla.Web/Services/SettingsClient.cs
using System.Net.Http.Json;
using System.Text.Json;
using Kuulla.Web.Models;
using Kuulla.Web.Services.Sync;

namespace Kuulla.Web.Services;

public class SettingsClient(KuullaApiClient apiClient)
{
    private const string DeviceId = "web";

    private static readonly JsonSerializerOptions JsonOptions = new(JsonSerializerDefaults.Web);

    public async Task<UserSettings> GetSettingsAsync(CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.GetAsync("api/settings", cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateUnlistenedEpisodeCountAsync(
        UnlistenedEpisodeCount unlistenedEpisodeCount, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings", new { UnlistenedEpisodeCount = unlistenedEpisodeCount }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateSubscriptionSortOrderAsync(
        SubscriptionSortOrder subscriptionSortOrder, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/subscription-sort-order",
            new { SubscriptionSortOrder = subscriptionSortOrder }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateSubscriptionManualOrderAsync(
        IReadOnlyList<string> showIds, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/subscription-manual-order", new { ShowIds = showIds }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateHideCaughtUpShowsAsync(
        bool hideCaughtUpShows, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/hide-caught-up-shows",
            new { HideCaughtUpShows = hideCaughtUpShows }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> GetShowSettingsAsync(string showId, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.GetAsync($"api/settings/shows/{Uri.EscapeDataString(showId)}", cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> UpdateShowUnlistenedEpisodeCountAsync(
        string showId, UnlistenedEpisodeCount? unlistenedEpisodeCount, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            $"api/settings/shows/{Uri.EscapeDataString(showId)}",
            new { UnlistenedEpisodeCount = unlistenedEpisodeCount }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateAutoArchiveRuleAsync(
        AutoArchiveRule autoArchiveRule, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/auto-archive", new { AutoArchiveRule = autoArchiveRule }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> UpdateShowAutoArchiveRuleAsync(
        string showId, AutoArchiveRule? autoArchiveRule, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            $"api/settings/shows/{Uri.EscapeDataString(showId)}/auto-archive",
            new { AutoArchiveRule = autoArchiveRule }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> UpdateShowAutoSkipAsync(
        string showId, int? autoSkipIntroSeconds, int? autoSkipOutroSeconds,
        CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            $"api/settings/shows/{Uri.EscapeDataString(showId)}/auto-skip",
            new { AutoSkipIntroSeconds = autoSkipIntroSeconds, AutoSkipOutroSeconds = autoSkipOutroSeconds },
            JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> UpdateShowPlaybackSpeedAsync(
        string showId, float? playbackSpeed, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            $"api/settings/shows/{Uri.EscapeDataString(showId)}/playback-speed",
            new { PlaybackSpeed = playbackSpeed }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> UpdateShowSmartSpeedAsync(
        string showId, bool? smartSpeed, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            $"api/settings/shows/{Uri.EscapeDataString(showId)}/smart-speed",
            new { SmartSpeed = smartSpeed }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> UpdateShowNotificationsEnabledAsync(
        string showId, bool? notificationsEnabled, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            $"api/settings/shows/{Uri.EscapeDataString(showId)}/notifications",
            new { NotificationsEnabled = notificationsEnabled }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateAutoSkipAsync(
        int autoSkipIntroSeconds, int autoSkipOutroSeconds, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/auto-skip",
            new { AutoSkipIntroSeconds = autoSkipIntroSeconds, AutoSkipOutroSeconds = autoSkipOutroSeconds },
            JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdatePlaybackSpeedAsync(
        float playbackSpeed, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/playback-speed", new { PlaybackSpeed = playbackSpeed }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateAutoDeleteRuleAsync(
        AutoDeleteRule autoDeleteRule, int autoDeleteAfterDays, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/auto-delete",
            new { AutoDeleteRule = autoDeleteRule, AutoDeleteAfterDays = autoDeleteAfterDays },
            JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> UpdateShowAutoDeleteRuleAsync(
        string showId, AutoDeleteRule? autoDeleteRule, int? autoDeleteAfterDays, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            $"api/settings/shows/{Uri.EscapeDataString(showId)}/auto-delete",
            new { AutoDeleteRule = autoDeleteRule, AutoDeleteAfterDays = autoDeleteAfterDays },
            JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateAutoDownloadNewEpisodesAsync(
        bool autoDownloadNewEpisodes, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/auto-download", new { AutoDownloadNewEpisodes = autoDownloadNewEpisodes }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> UpdateShowAutoDownloadNewEpisodesAsync(
        string showId, bool? autoDownloadNewEpisodes, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            $"api/settings/shows/{Uri.EscapeDataString(showId)}/auto-download",
            new { AutoDownloadNewEpisodes = autoDownloadNewEpisodes }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateAutoAddNewEpisodesToUpNextAsync(
        bool autoAddNewEpisodesToUpNext, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/auto-add-up-next",
            new { AutoAddNewEpisodesToUpNext = autoAddNewEpisodesToUpNext }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> UpdateShowAutoAddNewEpisodesToUpNextAsync(
        string showId, bool? autoAddNewEpisodesToUpNext, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            $"api/settings/shows/{Uri.EscapeDataString(showId)}/auto-add-up-next",
            new { AutoAddNewEpisodesToUpNext = autoAddNewEpisodesToUpNext }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateUpNextInsertPositionAsync(
        UpNextInsertPosition upNextInsertPosition, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/up-next-insert-position",
            new { UpNextInsertPosition = upNextInsertPosition }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<ShowSettings> UpdateShowUpNextInsertPositionAsync(
        string showId, UpNextInsertPosition? upNextInsertPosition, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            $"api/settings/shows/{Uri.EscapeDataString(showId)}/up-next-insert-position",
            new { UpNextInsertPosition = upNextInsertPosition }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<ShowSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateSmartSpeedAsync(
        bool smartSpeed, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/smart-speed", new { SmartSpeed = smartSpeed }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    public async Task<UserSettings> UpdateSleepTimerDefaultDurationAsync(
        int sleepTimerDefaultDurationMinutes, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var response = await client.PutAsJsonAsync(
            "api/settings/sleep-timer-default-duration",
            new { SleepTimerDefaultDurationMinutes = sleepTimerDefaultDurationMinutes }, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        return (await response.Content.ReadFromJsonAsync<UserSettings>(JsonOptions, cancellationToken))!;
    }

    // Polls (empty changes) or pushes (one change) via POST /api/sync/settings, mirroring
    // EpisodeStateClient.SyncAsync — see docs/sync-conventions.md.
    public async Task<SyncCheckResult<UserSettings>> SyncAsync(
        string localHash, DateTimeOffset lastSyncedAt, CancellationToken cancellationToken = default)
    {
        var client = await apiClient.CreateClientAsync();
        var body = new
        {
            DeviceId,
            LastSyncedAt = lastSyncedAt,
            LocalHash = localHash,
            Changes = Array.Empty<object>(),
        };
        var response = await client.PostAsJsonAsync("api/sync/settings", body, JsonOptions, cancellationToken);
        response.EnsureSuccessStatusCode();
        // The API's SyncSettingsResult and SyncCheckResult<T> already share the same
        // (ServerChanges, SyncedAt, Hash) shape, so this deserializes straight into it rather
        // than through a redundant private mirror record.
        return (await response.Content.ReadFromJsonAsync<SyncCheckResult<UserSettings>>(JsonOptions, cancellationToken))!;
    }
}
